File: trikot-analytics/firebase-ktx/src/main/java/com/mirego/trikot/analytics/FirebaseAnalyticsService.kt
package com.mirego.trikot.analytics

import android.content.Context
import android.os.Bundle
import com.google.firebase.analytics.FirebaseAnalytics
import com.mirego.trikot.streams.reactive.promise.Promise

class FirebaseAnalyticsService(context: Context, analyticsEnabled: Boolean = true) :
    AnalyticsService {

    private var firebaseAnalytics = FirebaseAnalytics.getInstance(context)

    private var analyticsEnabled: Boolean = analyticsEnabled

    override var isEnabled: Boolean
        get() = analyticsEnabled
        set(value) {
            analyticsEnabled = value
            firebaseAnalytics.setAnalyticsCollectionEnabled(value)
        }

    override val name: String = "FirebaseAnalytics"

    private var superProperties = emptyMap<String, Any>()

    override fun distinctAppId() = Promise.create<String> { resolve, reject ->
        firebaseAnalytics.appInstanceId
            .addOnCompleteListener { task ->
                val result = task.result
                if (task.isSuccessful && result != null) {
                    resolve(result)
                } else if(!task.isSuccessful) {
                    reject(Throwable(task.exception))
                } else {
                    reject(Throwable("Empty result"))
                }
            }
            .addOnCanceledListener {
                reject(Throwable("Canceled"))
            }
    }

    override fun identifyUser(userId: String, properties: AnalyticsPropertiesType) {
        firebaseAnalytics.setUserId(userId)
        properties.forEach {
            firebaseAnalytics.setUserProperty(it.key, it.value.toString())
        }
    }

    // This functionality isn't supported with Firebase Analytics
    override fun incrementUserProperties(incrementalProperties: AnalyticsIncrementalProperties) =
        Unit

    override fun logout() {
        firebaseAnalytics.setUserId(null)
    }

    override fun setSuperProperties(properties: AnalyticsPropertiesType) {
        superProperties = superProperties + properties
    }

    override fun setUserProperties(properties: AnalyticsPropertiesType) {
        properties.forEach {
            firebaseAnalytics.setUserProperty(it.key, it.value.toString())
        }
    }

    override fun unsetSuperProperties(propertyNames: List<String>) {
        superProperties = superProperties.entries
            .filter { !propertyNames.contains(it.key) }
            .fold(emptyMap()) { acc, value ->
                acc + value.toPair()
            }
    }

    override fun unsetAllSuperProperties() {
        superProperties = emptyMap()
    }

    override fun trackEvent(event: AnalyticsEvent, properties: AnalyticsPropertiesType) {
        val allProperties = properties + superProperties
        val bundle = Bundle()
        allProperties.forEach {
            bundle.putString(it.key, it.value.toString())
        }
        firebaseAnalytics.logEvent(event.name, bundle)
    }
}

File: trikot-analytics/swift-extensions/firebase/FirebaseAnalyticsService.swift
import FirebaseAnalytics
import Foundation
import TRIKOT_FRAMEWORK_NAME

public class FirebaseAnalyticsService: AnalyticsService {
    public var name: String = "FirebaseAnalytics"
    private var superProperties = [String: Any]()

    public init(enableAnalytics: Bool = true) {
        isEnabled = enableAnalytics
    }

    public var isEnabled: Bool {
        didSet {
            Analytics.setAnalyticsCollectionEnabled(isEnabled)
        }
    }

    public func distinctAppId() -> Promise<NSString>  {
        PromiseCompanion().resolve(value: Analytics.appInstanceID) as! Promise<NSString>
    }

    public func identifyUser(userId: String, properties: [String: Any]) {
        Analytics.setUserID(userId)
        properties.forEach { Analytics.setUserProperty(anyToString($0.value), forName: $0.key) }
    }

    public func incrementUserProperties(incrementalProperties: [String: Any]) {
        // This functionality isn't supported with Firebase Analytics
    }

    public func logout() {
        Analytics.setUserID(nil)
    }

    public func setSuperProperties(properties: [String: Any]) {
        properties.forEach { superProperties[$0.key] = $0.value }
    }

    public func setUserProperties(properties: [String: Any]) {
        properties.forEach { Analytics.setUserProperty(anyToString($0.value), forName: $0.key) }
    }

    public func trackEvent(event: AnalyticsEvent, properties: [String: Any]) {
        var allProperties = [String: Any]()
        properties.forEach { allProperties[$0.key] = $0.value }
        superProperties.forEach { allProperties[$0.key] = $0.value }
        Analytics.logEvent(event.name, parameters: allProperties)
    }

    public func unsetAllSuperProperties() {
        superProperties.removeAll()
    }

    public func unsetSuperProperties(propertyNames: [String]) {
        propertyNames.forEach { superProperties.removeValue(forKey: $0) }
    }

    private func anyToString(_ value: Any) -> String? {
        if let value = value as? String {
            return value
        } else {
            return "\(value)"
        }
    }
}
